otherlibs/stdlib_alpha/capsule.ml

```diff
@@ -376,56 +653,90 @@ module Rwlock = struct
+      match t.state with
+      | Poisoned -> Rw.unlock t.rwlock; raise Poisoned
+      | Frozen -> Rw.unlock t.rwlock; raise Frozen
+      | Open ->
+        let pw : k Password.t = Password.unsafe_mk () in
+        match f pw with
         | x -> Rw.unlock t.rwlock; x
         | exception exn ->
-          t.poisoned <- true;
-          Rw.unlock t.rwlock;
           let exn =
             match exn with
-            | Encapsulated (name, data) ->
-              (match Name.equality_witness name t.name with
+            | Encapsulated (id, data) ->
+              (match Password.Id.equality_witness (Password.id pw) id with
+               | Some Equal -> Data.unsafe_get data
+               | None -> exn)
+            | Encapsulated_shared (id, data) ->
+              (match Password.Id.equality_witness (Password.id pw) id with
                | Some Equal -> Data.unsafe_get data
                | None -> exn)
             | _ -> exn
           in
-          reraise exn
+          Rw.unlock t.rwlock;
+          t.state <- Poisoned;
+          raise exn
 
   let[@inline] with_read_lock :
-    'k t
-    -> ('k Password.Shared.t @ local -> 'a) @ local
+    type k.
+    k t
+    -> (k Password.Shared.t @ local -> 'a) @ local
     -> 'a
     @@ portable
     = fun t f ->
       Rw.lock_read t.rwlock;
-      match t.poisoned with
-      | true -> Rw.unlock t.rwlock; reraise Poisoned
-      | false ->
-        match f (Password.Shared.unsafe_mk t.name) with
+      match t.state with
+      | Poisoned -> Rw.unlock t.rwlock; raise Poisoned
+      | Open | Frozen ->
+        let pw : k Password.Shared.t = Password.Shared.unsafe_mk () in
+        match f pw with
         | x -> Rw.unlock t.rwlock; x
         | exception exn ->
-          (* Here we are not poisoning the RwLock, see [capsule.mli] for explanation *)
+          let exn =
+            match exn with
+            | Encapsulated_shared (id, data) ->
+              (match Password.Id.equality_witness (Password.Shared.id pw) id with
+               | Some Equal -> Data.unsafe_get data
+               | None -> exn)
+            | _ -> exn
+          in
+          (* This racy write is ok according to the memory model, because:
+             1. All threads which race to write here are writing [Frozen],
+                and the only other write to this field in the program was
+                the initial write setting it to [Open].
+             2. All threads which race to read here do not distinguish between
+                [Open] and [Frozen].
+             All operations that distinguish between [Open] and [Frozen]
+             are protected by the write lock. *)
+          t.state <- Frozen;
           Rw.unlock t.rwlock;
-          reraise exn
+          raise exn
+
+  let[@inline] freeze t =
+    Rw.lock_write t.rwlock;
+    match t.state with
+    | Poisoned -> raise Poisoned
+    | Open | Frozen ->
+      t.state <- Frozen;
+      Rw.unlock t.rwlock;
+      Key.unsafe_mk ()
```

Comment:
In `with_read_lock` above, the `t.state <- Frozen` write is performed with only read lock.  Shouldn't it be safe to do the write here also under only read lock?

Reply:
I agree. This should take the read lock rather than the write lock.